Add basic auth types from gateway

burtonr opened the pull request:
This changeset adds basic auth types and functions originally only
in the gateway. This will allow providers to perform authorization
checks independently. In cases like Swarm, where the functions are
accessible directly, this is needed to ensure calls are authorized

Signed-off-by: Burton Rheutan <rheutan7@gmail.com>

I also took the liberty of updating the copyright lines to ensure consistency across the projects.

This change is required in order to complete the issues in faas and faas-swarm:
https://github.com/openfaas/faas/issues/836
https://github.com/openfaas/faas-swarm/issues/31

alexellis commented on types/requests.go:
```diff
@@ -1,8 +1,13 @@
-// Copyright (c) Alex Ellis 2017. All rights reserved.
```
Should never remove a personal copyright from a file.

alexellis commented on types/basic_auth.go:
```diff
@@ -0,0 +1,54 @@
+// Copyright (c) OpenFaaS Project. All rights reserved.
```
This should be `OpenFaaS Author(s)` unless it had a different copyright in its original source file?

alexellis commented on types/requests.go:
```diff
@@ -20,3 +25,21 @@ type ProviderVersion struct {
 }
+
+// DecorateWithBasicAuth enforces basic auth as a middleware with given credentials
+func DecorateWithBasicAuth(next http.HandlerFunc, credentials *BasicAuthCredentials) http.HandlerFunc {
```
I think this might be better in a `handlers` package rather than `types`.

burtonr commented:
@alexellis I've updated the commit. 

Added the "OpenFaaS Author(s)" to the copyright, without removing your name. 
I've also added a "handlers" package that contains the `DecorateWithBasicAuth` function. I renamed the `types` file to `credentials` as I felt that was a bit more descriptive than the previous `basic_auth` file name. Especially since the handler is named `basic_auth` as well.

Let me know what you think!

rgee0 commented on handlers/basic_auth.go:
```diff
@@ -0,0 +1,28 @@
+		user, password, ok := r.BasicAuth()
+		w.Header().Set("WWW-Authenticate", `Basic realm="Restricted"`)
+
+		if !ok || !(credentials.Password == password && credentials.User == user) {
```
This is probably a personal thing but I find this a bit mind bending.

Id find this easier to read:

```
if !ok || credentials.Password != password || credentials.User != user {
```

rgee0 commented on types/credentials.go:
```diff
@@ -0,0 +1,54 @@
+		return nil, fmt.Errorf("invalid SecretMountPath specified for reading secrets")
+	}
+
+	userPath := path.Join(r.SecretMountPath, "basic-auth-user")
```
Nice to see

alexellis replied:
I believe this is a verbatim lift/shift from the faas project @rgee0 ?

rgee0 commented on types/credentials.go:
```diff
@@ -0,0 +1,54 @@
+		return nil, fmt.Errorf("Unable to load %s", userPassword)
+	}
+
+	credentials = &BasicAuthCredentials{
```
Should we handle potential new lines here? Genuine question as I'm 50/50 on it but we have had scenarios where secret creation has gone awry due to various methods implicitly adding \r

burtonr commented:
@rgee0 and @stefanprodan, thanks for the suggestions! @alexellis I've made the updates suggested making the credential `if` check clearer and remove newline characters from the username and password strings. Let me know what you think.

alexellis commented:
cc @bartsmykla @viveksyngh - both of you were interested in reading credentials. cc @s8sg since we do this in OpenFaaS Cloud too.

alexellis commented:
I believe I've completed the changes for this in the latest release and will now attempt to vendor them back into the faas project.

alexellis commented:
@burtonr thanks for taking an initial stab at this.
